src/traces/sankey/defaults.js

```diff
@@ -0,0 +1,136 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var Lib = require('../../lib');
+var attributes = require('./attributes');
+var d3 = require('d3');
+
+
+function linksDefaults(traceIn, traceOut) {
+    var linksIn = traceIn.links || [],
+        linksOut = traceOut.links = [];
+
+    var linkIn, linkOut, i;
+
+    function coerce(attr, dflt) {
+        return Lib.coerce(linkIn, linkOut, attributes.links, attr, dflt);
+    }
+
+    for(i = 0; i < linksIn.length; i++) {
+        linkIn = linksIn[i];
+        linkOut = {};
+
+        if(!Lib.isPlainObject(linkIn)) {
+            continue;
+        }
+
+        var visible = coerce('visible');
+
+        if(visible) {
+            coerce('label');
+            coerce('value');
+            coerce('source');
+            coerce('target');
+            coerce('color');
+        }
+
+        linkOut._index = i;
+        linksOut.push(linkOut);
+    }
+}
+
+function nodesDefaults(traceIn, traceOut) {
+    var nodesIn = traceIn.nodes || [],
+        nodesOut = traceOut.nodes = [];
+
+    var nodeIn, nodeOut, i, j, link, foundUse, visible,
+        usedNodeCount = 0,
+        indexMap = [];
+
+    var defaultPalette = d3.scale.category20();
+
+    function coerce(attr, dflt) {
+        return Lib.coerce(nodeIn, nodeOut, attributes.nodes, attr, dflt);
+    }
+
+    for(i = 0; i < nodesIn.length; i++) {
+        nodeIn = nodesIn[i];
+
+
+
+        foundUse = false;
+        for(j = 0; j < traceOut.links.length && !foundUse; j++) {
```

Comment:
Just for my education, since as you say, defaults must loop through it anyway (tho not nested) would putting it in `calc` result in more efficient code, or are you referring to the convention where computational things are usually in `calc`? I put it in `defaults` without much thought simply b/c filtering of unrendered elements is typically done in `defaults` and didn't think of it as 'calculation'.